### PR TITLE
Fixed cuda grads for interpolate::trilinear on non-contig grad output

### LINKED ISSUE #113642: interpolate::trilinear returns wrong gradients on CUDA
### 🐛 Describe the bug

```python
import torch
import numpy as np

grad_output_ncthw = torch.tensor([[[[[1]]], [[[1]]]]], dtype = torch.float32).cuda()
temp = torch.tensor([[[[[1, 1]]]]], dtype = torch.float32).cuda()
grad_output_permute2ncthw = temp.permute(0, 4, 1, 2, 3).contiguous()
print("equal:", grad_output_ncthw == grad_output_permute2ncthw)
print("1st:", grad_output_ncthw.shape, grad_output_ncthw.stride(), grad_output_ncthw.is_contiguous())
print("2nd:", grad_output_permute2ncthw.shape, grad_output_permute2ncthw.stride(), grad_output_permute2ncthw.is_contiguous())

use_1st = False
if use_1st:
    print("use 1st")
    grad_output = grad_output_ncthw
else:
    print("use 2nd")
    grad_output = grad_output_permute2ncthw

interp_input = torch.randn([1, 2, 4, 8, 16], dtype = torch.float32).cuda()

interp_input.requires_grad = True

interp_output = torch.nn.functional.interpolate(interp_input, scale_factor = [0.25, 0.125, 0.0625], mode = "trilinear", align_corners = False, recompute_scale_factor = False)

interp_output.backward(grad_output)
torch.set_printoptions(threshold = 200000)
grad_x = interp_input.grad
print(grad_x.shape, grad_x.stride())
print("grad_input", grad_x)
```

============ use 1st (desired result) =========
```
equal: tensor([[[[[True]]],


         [[[True]]]]], device='cuda:0')
1st: torch.Size([1, 2, 1, 1, 1]) (2, 1, 1, 1, 1) True
2nd: torch.Size([1, 2, 1, 1, 1]) (2, 1, 2, 2, 2) True
use 1st
torch.Size([1, 2, 4, 8, 16]) (1024, 512, 128, 16, 1)
grad_input tensor([[[[[0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000]],
...
```
=========== use 2nd (unexpected result) ===========
```
equal: tensor([[[[[True]]],


         [[[True]]]]], device='cuda:0')
1st: torch.Size([1, 2, 1, 1, 1]) (2, 1, 1, 1, 1) True
2nd: torch.Size([1, 2, 1, 1, 1]) (2, 1, 2, 2, 2) True
use 2nd
torch.Size([1, 2, 4, 8, 16]) (1024, 512, 128, 16, 1)
grad_input tensor([[[[[0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.1250, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.1250, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
           [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
            0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000]],
...
```
From what I see, `use_1st` and `use_2nd` only differ in stride values, but they are contiguous, so I expect both get the same right results.
I don't observe this kind of behavior on CPU.

### Versions

Collecting environment information...
PyTorch version: 2.2.0a0+gitfe5d885
Is debug build: False
CUDA used to build PyTorch: 12.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.31

Python version: 3.8.18 (default, Sep 11 2023, 13:40:15)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-150-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-PCIE-40GB
Nvidia driver version: 535.104.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          48
On-line CPU(s) list:             0-47
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           49
Model name:                      AMD EPYC 7402P 24-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         1496.762
CPU max MHz:                     2800.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        5589.71
Virtualization:                  AMD-V
L1d cache:                       768 KiB
L1i cache:                       768 KiB
L2 cache:                        12 MiB
L3 cache:                        128 MiB
NUMA node0 CPU(s):               0-47
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate sme ssbd mba sev ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] optree==0.10.0
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.2.0a0+gitfe5d885
[conda] blas                      1.0                         mkl  
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-include               2023.1.0         h06a4308_46344  
[conda] mkl-service               2.4.0            py38h5eee18b_1  
[conda] mkl_fft                   1.3.8            py38h5eee18b_0  
[conda] mkl_random                1.2.4            py38hdb19cb5_0  
[conda] numpy                     1.24.3           py38hf6e8229_1  
[conda] numpy-base                1.24.3           py38h060ed82_1  
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+6e4932cda8          pypi_0    pypi
[conda] torch                     2.2.0a0+gitfe5d885           dev_0    <develop>

cc @ezyang @gchanan @zou3519 @kadeng @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @ptrblck

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #113642

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -10007,31 +10007,41 @@ def test_upsamplingBicubic2d_aa_correctness(self, device, memory_format):
         self.assertEqual(expected_out, t_out)
 
     @parametrize_test("align_corners", [True, False])
-    def test_upsamplingTrilinear3d(self, device, align_corners):
+    @parametrize_test("memory_format", [torch.contiguous_format, torch.channels_last_3d])
+    def test_upsamplingTrilinear3d(self, device, align_corners, memory_format):
```
Comment (lezcano) [inline test/test_nn.py:10011]: Don't we have an opinfo for this?

File: aten/src/ATen/native/cuda/UpSampleTrilinear3d.cu [hunk 2]
```diff
@@ -363,6 +364,10 @@ static void upsample_trilinear3d_backward_out_cuda_template(
                 odata,
                 idata_ptr);
         C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+        if (!grad_input_.is_contiguous()) {
+            grad_input_.copy_(grad_input);
+        }
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/UpSampleTrilinear3d.cu:370]: This is visible from the outside, which is not correct.
Why do we need it at all?

Comment (vfdev-5) [reply aten/src/ATen/native/cuda/UpSampleTrilinear3d.cu:370]: This is what we typically do for backward ops I would say:
https://github.com/pytorch/pytorch/blob/41ffea2f99b91237ed2952c8d94c6b0e56548384/aten/src/ATen/native/cuda/UpSampleBilinear2d.cu#L464-L466

> Why do we need it at all?

grads algorithm only working for contiguous tensors and previously Line 324 is wrong when trying to make grad_input to be contiguous:
```
// A contiguous tensor is required for the kernel launch config
grad_input.contiguous();
```

Comment (lezcano) [reply aten/src/ATen/native/cuda/UpSampleTrilinear3d.cu:370]: Yes, I understand the bug, and the fix LGTM, but I don't think we should change the values of an input tensor in-place. That will surely break double gradients though that variable.

File: test/test_nn.py (hunk 1 above)
Comment (vfdev-5) [reply test/test_nn.py:10011]: Yes, there is one, but it does not test the failed case. Also `test_upsamplingTrilinear3d` already existing, I just put memory_format into parametrization and added grads consistency check between CF and CL which testing the fix proposed in this PR

File: aten/src/ATen/native/cuda/UpSampleTrilinear3d.cu (hunk 2 above)
Comment (lezcano) [reply aten/src/ATen/native/cuda/UpSampleTrilinear3d.cu:370]: nvm, this is completely fine

File: test/test_nn.py (hunk 1 above)
Comment (lezcano) [reply test/test_nn.py:10011]: It'd be nice to add the relevant failing samples to the OpInfo, that way we would get proper testing for these inputs.

**Review by lezcano — APPROVED** [review]

File: test/test_nn.py (hunk 1 above)
Comment (vfdev-5) [reply test/test_nn.py:10011]: Actually, it is not about specific sample but a specific way of testing. We have to ensure that grad_input is consistent for grad_output contig vs grad_output channels_last3d. Probably, there is no such test in test_ops.py and gradcheck does not do that...

Comment (lezcano) [reply test/test_nn.py:10011]: we do have a non contiguous test, I think that would exercise this path, wouldn't it?

Comment (lezcano) [reply test/test_nn.py:10011]: ah, no, this is for the backward. Ok, let's leave it as is

Comment (lezcano) [reply test/test_nn.py:10011]: hmmm, the non contig test does exercise the backward as well tho... oh well

**Pull request closed without merging**